xrmocap/transform/keypoints3d/optim/rm_duplicate.py

```diff
@@ -0,0 +1,143 @@
+# yapf: disable
+import logging
+import numpy as np
+import torch
+from typing import Union
+
+from xrmocap.data_structure.keypoints import Keypoints
+from xrmocap.ops.top_down_association.identity_tracking.builder import (
+    BaseTracking, build_identity_tracking,
+)
+from .base_optimizer import BaseOptimizer
+
+# yapf: enable
+
+
+class RemoveDuplicate(BaseOptimizer):
+    """This optimization tool helps remove duplicate identities by the L2
+    distance between 3D keypoints for each person and add tracked identities to
+    optimized 3D keypoints."""
+
+    def __init__(
+        self,
+        verbose: bool = True,
+        threshold: float = 2.0,
+        keep: str = 'by_index',
+        identity_tracking: Union[None, dict, BaseTracking] = None,
+        logger: Union[None, str, logging.Logger] = None,
+    ):
+        """Initialization for RemoveDuplicate optimizor.
+
+        Args:
+            verbose (bool, optional):
+                Whether to log info. Defaults to True.
+            threshold (float, optional):
+                Threshold for duplicate 3D keypoints. Defaults to 2.0.
+            keep (str, optional):
+                Mode to keep 3D keypoints, 'by_index' keeps the lowest index,
+                'by_conf' keeps the highest confidence. Defaults to 'by_index'.
+            logger (Union[None, str, logging.Logger], optional):
+                Logger for logging. If None, root logger will be selected.
+                Defaults to None.
+        """
+
+        super().__init__(verbose=verbose, logger=logger)
+        self.threshold = threshold
+        self.keep = keep
+
+        if isinstance(identity_tracking, dict):
+            identity_tracking['logger'] = logger
+            self.identity_tracking = build_identity_tracking(identity_tracking)
+        else:
+            self.identity_tracking = identity_tracking
+
+    def optimize_keypoints3d(self, keypoints3d: Keypoints,
+                             **kwargs: dict) -> Keypoints:
+        """Forward function for RemoveDuplicate optimizor.
+
+        Args:
+            keypoints3d (Keypoints):
+                Keypoints3d to be optimized.
+
+        Returns:
+            Keypoints: Optimized keypoints3d.
+        """
+        # get kps3d array
+        keypoints3d_optim = keypoints3d.clone()
+        kps3d = keypoints3d_optim.get_keypoints()
+        n_frame, n_max_person, n_kps, _ = kps3d.shape
+
+        kps3d_optim = np.full((n_frame, n_max_person, n_kps, 4), np.nan)
+
+        for frame_idx in range(n_frame):
+            kps3d_frame = kps3d[frame_idx, ...]
+            kps3d_frame = kps3d_frame[~np.isnan(kps3d_frame[:, 0, 0])]
+
+            # calculate distance and remove duplicate
+            dist = self.get_kps3d_dist(kps3d_frame)
```

Comment:
Is it possible to do this in batch manner? As [torch.cdist](https://pytorch.org/docs/stable/generated/torch.cdist.html) supports batch operation.

Reply:
Here we remove the `nan` value in each frame and only calculate the dist matrix for valid persons in the frame, if we do it in batch outside the loop for all the frames, the dist matrix might include a lot of `nan` values for those frames with only a few persons.